content/blog/k8s-122-challenge/index.md

```diff
@@ -0,0 +1,51 @@
+
+### Including the etcd and NATS clients
+We wanted to use the etcd-cluster and nats-cluster clients to detect old NATS and etcd dependencies. We hit a [strange compilation error](https://github.com/coreos/etcd-operator/issues/2167) while attempting to include the clients. It turns out that our packaged version of client-go is incompatible with the packaged version in the clients. The client-go developers introduced a [breaking interface change](https://github.com/kubernetes/client-go/commit/ae9f6b2601c8f8b97ad2865943f423d65539ffdb) and Go’s dependency management disallows [two different incompatible versions](https://research.swtch.com/vgo-import#dependency_story) with the same import path.
+
+Fortunately, we had recently used `client-gen` [to generate client code for Vizier CustomResources](https://github.com/pixie-io/pixie/commit/3950a03d723029b4280e5267f0c2c9c78f2dd7c8). We decided to do the same with nats-cluster and etcd-cluster, [vendoring the clients into our operator code](https://github.com/pixie-io/pixie/search?q=generate&type=commits). Once you have a hammer, everything starts to look like a nail.
+
+Now during our updates, the Vizier operator simply uses the vendored clients to check for etcd-cluster or nats-cluster resources deployed by old versions of Vizier and replaces them with the new StatefulSet versions if they happen to exist. Future Viziers will only deploy with the StatefulSet Versions.
+
+## Conclusion
+The removal of the beta CRD API from Kubernetes 1.22 posed a unique challenge for our team. Switching to StatefulSet NATS and etcd removed the deprecated third-party operator dependencies. Updating all of our running instances allowed us to avoid bifurcating our deployments.
+
+If you’re looking to use StatefulSet NATS and etcd, check out our [NATS](https://github.com/pixie-io/pixie/tree/main/k8s/vizier_deps/base/nats) and [etcd](https://github.com/pixie-io/pixie/tree/main/k8s/vizier_deps/base/etcd) yamls as references. You should also check our [blog on how etcd works ](https://blog.px.dev/etcd-6-tips/)and [why we chose NATS as our message bus](https://blog.px.dev/hybrid-architecture/#choosing-a-message-bus).
+
+Let us know if you found this type of post interesting! We’re trying to not only open-source our codebase, but also openly discuss the challenges we’ve faced and lessons that we’ve learned along the way.
+
+_Find us on [Slack](https://slackin.px.dev/), [GitHub](https://github.com/pixie-io/pixie/blob/main/CONTRIBUTING.md), or Twitter at [@pixie_run](https://twitter.com/pixie_run)._
+
+[^1]: the `nats-operator` code for managing CRDs has since[ been updated to the GA API ](https://github.com/nats-io/nats-operator/pull/333), but they have not released a new version.
```

Comment:
```suggestion
## Footnote
[^1]: The `nats-operator` code for managing CRDs has since[ been updated to the GA API ](https://github.com/nats-io/nats-operator/pull/333), but they have not released a new version.
```